code/controllers/subsystem/processing/overlays.dm

```diff
@@ -0,0 +1,115 @@
+var/datum/subsystem/processing/overlays/SSoverlays
+
+/datum/subsystem/processing/overlays
+	name = "Overlays"
+	flags = SS_TICKER|SS_NO_INIT|SS_FIRE_IN_LOBBY
+	wait = 1
+	priority = 500
+	can_fire = FALSE
+
+	stat_tag = "Ov"
+	currentrun = null
+
+/datum/subsystem/processing/overlays/New()
+	NEW_SS_GLOBAL(SSoverlays)
+
+/datum/subsystem/processing/overlays/Recover()
+	processing = SSoverlays.processing
+
+/datum/subsystem/processing/overlays/fire()
+	while(processing.len)
+		var/atom/thing = processing[processing.len]
+		processing.len--
+		if(thing)
+			if(LAZYLEN(thing.priority_overlays) && LAZYLEN(thing.our_overlays))
+				thing.overlays = (thing.our_overlays | thing.priority_overlays)
+			else if(LAZYLEN(thing.our_overlays))
+				thing.overlays = thing.our_overlays
+			else if(LAZYLEN(thing.priority_overlays))			//Do these two have to be copied? We don't want assignments to them to trigger overlay updates. But, idk, bruh, BYOND. - Cyberboss
+				thing.overlays = thing.priority_overlays
+			else
+				LAZYCLEARLIST(thing.overlays)
+		if(MC_TICK_CHECK)
+			break
+	can_fire = processing.len > 0
+
+/atom/proc/cut_overlays(priority = FALSE)
+	var/list/cached_overlays = our_overlays
+	var/list/cached_priority = priority_overlays
+	
+	var/need_compile = FALSE
+
+	if(LAZYLEN(cached_overlays)) //don't queue empty lists, don't cut priority overlays
+		cached_overlays.Cut()  //clear regular overlays
+		need_compile = TRUE
+
+	if(priority && LAZYLEN(cached_priority))
+		cached_priority.Cut()
+		need_compile = TRUE
+
+	if(need_compile)
+		SSoverlays.processing[src] = src
```

Comment:
It would be better and have massively less overhead if this tracked rather or not an atom was in the processing list using a flag.

[src] = src has less overhead then |= but being able to just do += is fastest by a far bit.